protocols/ledger/ledger.go

```diff
@@ -32,39 +32,44 @@ func (l *LedgerManager) HandleRequest(ledger *channel.TwoPartyLedger, request pr
 	if err != nil {
 		return protocols.SideEffects{}, fmt.Errorf("error finding a supported state: %w", err)
 	}
-
-	asset := types.Address{} // todo: loop over request.amount's assets
 	nextState := supported.Clone()
+	nextState.Outcome = outcome.Exit{}
+	// TODO: We should iterate over assets in RightAmount as well. For now we rely on the assumption that LeftAmount and RightAmount will have the same assets.
+	for asset := range request.LeftAmount {
+
+		currentLeftAmount := supported.Outcome.TotalAllocatedFor(request.Left)[asset]
+		currentRightAmount := supported.Outcome.TotalAllocatedFor(request.Right)[asset]
+		// Calculate the amounts
+		leftAmount := big.NewInt(0).Sub(currentLeftAmount, request.LeftAmount[asset])
+		rightAmount := big.NewInt(0).Sub(currentRightAmount, request.RightAmount[asset])
+		total := big.NewInt(0).Add(request.LeftAmount[asset], request.RightAmount[asset])
+		if leftAmount.Cmp(big.NewInt(0)) < 0 {
+			return protocols.SideEffects{}, fmt.Errorf("Allocation for %x cannot afford the amount %d", request.Left, request.LeftAmount[asset])
+		}
+		if rightAmount.Cmp(big.NewInt(0)) < 0 {
+			return protocols.SideEffects{}, fmt.Errorf("Allocation for %x cannot afford the amount %d", request.Right, request.RightAmount[asset])
+		}
 
-	// Calculate the amounts
-	amountPerParticipant := big.NewInt(0).Div(request.Amount[asset], big.NewInt(2))
-	leftAmount := big.NewInt(0).Sub(nextState.Outcome.TotalAllocatedFor(request.Left)[asset], amountPerParticipant)
-	rightAmount := big.NewInt(0).Sub(nextState.Outcome.TotalAllocatedFor(request.Right)[asset], amountPerParticipant)
-	if leftAmount.Cmp(big.NewInt(0)) < 0 {
-		return protocols.SideEffects{}, fmt.Errorf("Allocation for %x cannot afford the amount %d", request.Left, amountPerParticipant)
+		newOutcome := outcome.SingleAssetExit{
+			Allocations: outcome.Allocations{
+				outcome.Allocation{
+					Destination: request.Left,
+					Amount:      leftAmount,
+				},
+				outcome.Allocation{
+					Destination: request.Right,
+					Amount:      rightAmount,
+				},
+				outcome.Allocation{
+					Destination:    request.Destination,
+					Amount:         total,
+					AllocationType: outcome.GuaranteeAllocationType,
+					Metadata:       guarantee,
+				},
+			}}
+
+		nextState.Outcome = append(nextState.Outcome, newOutcome)
```

Comment:
Don't we also need to make some deductions from existing entries in the outcome?

Reply:
Sorry I read it more carefully and looked at the test and I can see the deductions are there correctly. Variable names like `newLeftAmount` may have helped.

I'm still not sure about the `append`, though? Does that only work currently because there is no supported state on L35?

Reply:
This change has now been pulled out to #247 but I think you're right. Right now we could lose any exits for assets not in `leftAmount`.

Reply:
This has been pulled out to #247 so this discussion doesn't exactly apply to the current PR's code. That said in #247 it works by
- [Cloning](https://github.com/statechannels/go-nitro/blob/467a3efa30bfbf40f84b8a4f1d0c8f9a8ada46ad/protocols/ledger/ledger.go#L35) the state into `nextState`
- [Resetting](https://github.com/statechannels/go-nitro/blob/467a3efa30bfbf40f84b8a4f1d0c8f9a8ada46ad/protocols/ledger/ledger.go#L36) the outcome on `nextState`
- Creating new single asset exits based on the previous `supported.Outcome` and `ledgerRequest`

I think this does introduce a problem as we're only iterating through the assets of the ledger request. This means we would lose any `SingleAssetExit`s for assets not in the Ledger Request amounts.

Reply:
It looks like the "resetting outcome" step is missing on #247 right now? (it's on the commit you linked to, which apparently does not belong to a branch).

I agree about the problem of missing assets which are not in the ledger request. But I think there is an additional problem -- even if the asset _is_ in the ledger request, aren't we are going to lose existing allocations that don't allocate to `left` and `right`? For example, allocations which guarantee other virtual channels?

Reply:
We can move the discussion to #247